Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs

```diff
@@ -156,25 +155,56 @@ public void When_Join_Request_Encountered_Should_Find_Duplicate(string station1,
         public void CreateKeyMethod_Should_Produce_Expected_Key_For_JoinRequests()
         {
             // arrange
-            var joinPayload = (LoRaPayloadJoinRequest)this.joinRequest.Payload;
-            joinPayload.DevNonce = new DevNonce(0);
-            this.joinRequest.SetPayload(joinPayload);
+            this.joinPayload.DevNonce = new DevNonce(0);
+            this.joinRequest.SetPayload(this.joinPayload);
 
             var expectedKey = "60-DA-A3-A5-F7-DB-FA-20-0F-8C-82-84-0E-CF-5B-42-64-0B-70-F3-B7-21-8A-4C-6B-BD-67-DB-54-2E-75-A4";
 
             // act/assert
-            Assert.Equal(expectedKey, ConcentratorDeduplication.CreateCacheKey(this.joinRequest));
+            Assert.Equal(expectedKey, this.concentratorDeduplication.CreateCacheKey(this.joinPayload));
+        }
+
+        [Theory, MemberData(nameof(TestJoinRequestData.DataGenerator), MemberType = typeof(TestJoinRequestData))]
+        public void CreateKeyMethod_Should_Return_Expected_Keys_For_Different_JoinRequests(LoRaPayloadJoinRequest joinRequestPayload, string expectedKey)
+        {
+            Assert.Equal(expectedKey, this.concentratorDeduplication.CreateCacheKey(joinRequestPayload));
         }
         #endregion
 
-        [Fact]
-        public void CreateKeyMethod_Should_Throw_When_Used_With_Wrong_Type()
+        private static class TestPayloadData
```

Comment:
This seems like it can be simplified to very simple tests with inline data:

```c#
[Theory]
[InlineData("60daa3a5f7dbfa200f8c82840ecf5b42640b70f3b7218a4c6bbd67db542e75a4", 0, 0, 0)]
[InlineData("bf6781db771aef1c14559e2c22e7d1cf4f577777656e2cd6e3d41a6ea16a1786", 0x101010101010101UL, 0, 0)]
[InlineData("4025818beac7accd324e2a02ce15c89b72a481329d919ffe7a62d8ba3ac4e400", 0, 0x101010101010101UL, 0)]
[InlineData("b0ef53e122b3c40b579355219695430329f46c3b24936cbd734967780a609be2", 0, 0, 1)]
public void CreateCacheKey_JoinRequests(string expectedKey, ulong joinEui, ulong devEui, ushort devNonce)
{
    var payload = new LoRaPayloadJoinRequestLns(new MacHeader(MacMessageType.JoinRequest),
                                                new JoinEui(joinEui), new DevEui(devEui),
                                                new DevNonce(devNonce), new Mic(0));
#pragma warning disable CA1308 // Normalize strings to uppercase
    var key = this.concentratorDeduplication.CreateCacheKey(payload)
                                            .ToLowerInvariant()
                                            .Replace("-", string.Empty, StringComparison.Ordinal);
#pragma warning restore CA1308 // Normalize strings to uppercase
    Assert.Equal(expectedKey, key);
}
```

The benefits are:

- No need for separate data classes with byte arrays.
- Test data lives closer to or with the actual test method (inline).
- SHA hashes are shorter to read (though ideally `CreateCacheKey` methods can be modified to simply use `Hexadecimal` class to produce the hashes like that in the first place).
- Good to have expectations comes first, like in assertions.

In diff-speak, the changes would look as follows:

```diff
diff --git a/Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs b/Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs
index 9cbfbccd..00dd504f 100644
--- a/Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs
+++ b/Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs
@@ -164,10 +164,22 @@ namespace LoRaWan.Tests.Unit.NetworkServer
             Assert.Equal(expectedKey, this.concentratorDeduplication.CreateCacheKey(this.joinPayload));
         }
 
-        [Theory, MemberData(nameof(TestJoinRequestData.DataGenerator), MemberType = typeof(TestJoinRequestData))]
-        public void CreateKeyMethod_Should_Return_Expected_Keys_For_Different_JoinRequests(LoRaPayloadJoinRequest joinRequestPayload, string expectedKey)
+        [Theory]
+        [InlineData("60daa3a5f7dbfa200f8c82840ecf5b42640b70f3b7218a4c6bbd67db542e75a4", 0, 0, 0)]
+        [InlineData("bf6781db771aef1c14559e2c22e7d1cf4f577777656e2cd6e3d41a6ea16a1786", 0x101010101010101UL, 0, 0)]
+        [InlineData("4025818beac7accd324e2a02ce15c89b72a481329d919ffe7a62d8ba3ac4e400", 0, 0x101010101010101UL, 0)]
+        [InlineData("b0ef53e122b3c40b579355219695430329f46c3b24936cbd734967780a609be2", 0, 0, 1)]
+        public void CreateCacheKey_JoinRequests(string expectedKey, ulong joinEui, ulong devEui, ushort devNonce)
         {
-            Assert.Equal(expectedKey, this.concentratorDeduplication.CreateCacheKey(joinRequestPayload));
+            var payload = new LoRaPayloadJoinRequestLns(new MacHeader(MacMessageType.JoinRequest),
+                                                        new JoinEui(joinEui), new DevEui(devEui),
+                                                        new DevNonce(devNonce), new Mic(0));
+#pragma warning disable CA1308 // Normalize strings to uppercase
+            var key = this.concentratorDeduplication.CreateCacheKey(payload)
+                                                    .ToLowerInvariant()
+                                                    .Replace("-", string.Empty, StringComparison.Ordinal);
+#pragma warning restore CA1308 // Normalize strings to uppercase
+            Assert.Equal(expectedKey, key);
         }
         #endregion
 
@@ -191,22 +203,6 @@ namespace LoRaWan.Tests.Unit.NetworkServer
             };
         }
 
-        private static class TestJoinRequestData
-        {
-            private static readonly Memory<byte> BaseAppEui = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0 };
-            private static readonly Memory<byte> BaseDevEui = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0 };
-            private static readonly DevNonce BaseDevNonce = new DevNonce(0);
-
-            public static readonly TheoryData<LoRaPayloadJoinRequest, string> DataGenerator = new()
-            {
-                { new LoRaPayloadJoinRequest() { AppEUI = BaseAppEui, DevEUI = BaseDevEui, DevNonce = BaseDevNonce }, "60-DA-A3-A5-F7-DB-FA-20-0F-8C-82-84-0E-CF-5B-42-64-0B-70-F3-B7-21-8A-4C-6B-BD-67-DB-54-2E-75-A4" },
-                { new LoRaPayloadJoinRequest() { AppEUI = BaseAppEui, DevEUI = BaseDevEui, DevNonce = BaseDevNonce, Mhdr = new byte[] { 1 } }, "60-DA-A3-A5-F7-DB-FA-20-0F-8C-82-84-0E-CF-5B-42-64-0B-70-F3-B7-21-8A-4C-6B-BD-67-DB-54-2E-75-A4" },
-                { new LoRaPayloadJoinRequest() { AppEUI = new byte[] { 1, 1, 1, 1, 1, 1, 1, 1 }, DevEUI = BaseDevEui, DevNonce = BaseDevNonce }, "BF-67-81-DB-77-1A-EF-1C-14-55-9E-2C-22-E7-D1-CF-4F-57-77-77-65-6E-2C-D6-E3-D4-1A-6E-A1-6A-17-86" },
-                { new LoRaPayloadJoinRequest() { AppEUI = BaseAppEui, DevEUI = new byte[] { 1, 1, 1, 1, 1, 1, 1, 1 }, DevNonce = BaseDevNonce }, "40-25-81-8B-EA-C7-AC-CD-32-4E-2A-02-CE-15-C8-9B-72-A4-81-32-9D-91-9F-FE-7A-62-D8-BA-3A-C4-E4-00" },
-                { new LoRaPayloadJoinRequest() { AppEUI = BaseAppEui, DevEUI = BaseDevEui, DevNonce = new DevNonce(1) }, "B0-EF-53-E1-22-B3-C4-0B-57-93-55-21-96-95-43-03-29-F4-6C-3B-24-93-6C-BD-73-49-67-78-0A-60-9B-E2" },
-            };
-        }
-
         public void Dispose()
         {
             this.loRaDevice.Dispose();
```

I would recommend doing the same for the other test.

Reply:
Took the majority of your recommendations:
- Inline data (closer to test) without the need for a separate type (no extra mental overhead).
- Using new primitives instead of byte arrays (safer).
- Expectations in the beginning: debatable (I think most of our tests do not conform to this) but ok.

I did not take the recommendation about SHA as we would assert on something different than what the actual code uses (and also it's debatable whether it's more readable to have a continuous long string imo)

Reply:
Resolving due to inaction.